work_queue/src/work_queue.c

```diff
@@ -851,31 +847,76 @@ static int get_transfer_wait_time(struct work_queue *q, struct work_queue_worker
 	return timeout;
 }
 
-static void remove_factory(struct work_queue_factory_info *f)
+static int factory_trim_workers(struct work_queue *q, struct work_queue_factory_info *f)
+{
+	// Iterate through all workers and shut idle ones down
+	char *factory_name = f->name;
+	struct work_queue_worker *w;
+	char *key;
+	int i = 0;
+
+	hash_table_firstkey(q->worker_table);
+	while( f->connected_workers > f->max_workers &&
+			hash_table_nextkey(q->worker_table, &key, (void **) &w) ) {
+		if( !strcmp(factory_name, w->factory_name) &&
+			!itable_size(w->current_tasks) ) {
+			shut_down_worker(q, w);
+			i++;
+		}
+	}
+
+	// Trimmed some but not enough workers.
+	if (f->connected_workers > f->max_workers) f->trimmed = 1;
+	debug(D_WQ, "Trimmed %d workers from %s", i, factory_name);
+	return i;
+}
+
+static struct work_queue_factory_info* create_factory_info(struct work_queue *q, const char *name)
+{
+	struct work_queue_factory_info *f;
+	if ( (f = hash_table_lookup(q->factory_table, name)) ) return f;
+	if ( (f = malloc(sizeof(*f))) ) {
+		f->name = xxstrdup(name);
+		f->connected_workers = 0;
+		f->max_workers = INT_MAX;
+		f->trimmed = 0;
+		f->queried = 0;
+	} else {
+		debug(D_WQ, "Failed to allocate memory for factory info");
+	}
+	hash_table_insert(q->factory_table, name, f);
+	return f;
+}
+
+static void remove_factory_info(struct work_queue *q, const char *name)
 {
-	free(f->name);
-	free(f);
+	struct work_queue_factory_info *f = hash_table_lookup(q->factory_table, name);
+	if (f) {
+		free(f->name);
+		free(f);
+		hash_table_remove(q->factory_table, name);
+	} else {
+		debug(D_WQ, "Failed to remove unrecorded factory %s", name);
+	}
 }
 
 static void update_factory(struct work_queue *q, struct jx *j)
 {
 	const char *name = jx_lookup_string(j, "factory_name");
 	struct work_queue_factory_info *f;
-
 	if ( !(f = hash_table_lookup(q->factory_table, name)) ) {
-		// Create a new factory
-		debug(D_WQ, "new factory %s detected", name);
-		f = malloc(sizeof(*f));
-		if (!f) {
-			debug(D_NOTICE, "Cannot allocate memory for factory %s.", name);
-			return;
-		}
-		f->name = xxstrdup(name);
-		f->connected_workers = 0;
-		hash_table_insert(q->worker_table, f->name, f);
+		debug(D_WQ, "factory %s not recorded", name);
+		return;
 	}
 
-	return;
+	f->max_workers = jx_lookup_integer(j, "max_workers");
+	f->queried = 1;
+	// Trim workers if appropriate.
+	if (f->connected_workers > f->max_workers && !f->trimmed) {
```

Comment:
We can simplify to `if(!f->trimmed)`, as the other condition is checked in `factory_trim_workers`.